File: Book/Fortran/Fig_4.10_crit.f90
! sample compile command: 
!      gfortran -fopenmp -c Fig_4.10_crit.f90
! to generate *.o object file

program crit
   use omp_lib
   implicit none

   real :: res = 0.0
   integer :: niters = 1000
   real :: B
   integer :: i, id, nthrds

   interface
      function big_job(i)
         real :: big_job
         integer, intent(in) :: i
      end function big_job

      function consume(a)
         real :: consume
         real, intent(in) :: a
      end function consume
   end interface

   !$omp parallel private (id, nthrds)
      id = omp_get_thread_num()
      nthrds = omp_get_num_threads()
      do i = id, niters - 1, nthrds
         B = big_job(i)
         !$omp critical
            res = res + consume(B)
         !$omp end critical
      end do

   !$omp end parallel

end program crit

File: Book/C/Fig_4.10_crit.c
// sample compile command: "gcc -fopenmp -c Fig_4.10_crit.c" to generate *.o object file

#include <stdio.h>
#include <omp.h>

int main()
{
   float  res = 0.0;
   int niters = 1000;;
   #pragma omp parallel
   {     
      float B; int i, id, nthrds;
      id = omp_get_thread_num();
      nthrds = omp_get_num_threads();
      for (i = id; i < niters; i += nthrds) {
         B = big_job(i);
         #pragma omp critical
            res += consume(B);
      }
   } // end of parallel region
}
